lib/lru_redux/ttl/cache.rb

```diff
@@ -0,0 +1,192 @@
+        max_size, ttl = args
+
+        ttl ||= :none
+
+        raise ArgumentError.new(:max_size) if
+            max_size < 1
+        raise ArgumentError.new(:ttl) unless
+            ttl == :none || ((ttl.is_a? Numeric) && ttl >= 0)
+
+        @max_size = max_size
+        @ttl = ttl
+        @data_lru = {}
+        @data_ttl = {}
+      end
+
+      def max_size=(max_size)
+        max_size ||= @max_size
+
+        raise ArgumentError.new(:max_size) if
+            max_size < 1
+
+        @max_size = max_size
+
+        resize
+      end
+
+      def ttl=(ttl)
+        ttl ||= @ttl
+
+        raise ArgumentError.new(:ttl) unless
+            ttl == :none || ((ttl.is_a? Numeric) && ttl >= 0)
+
+        @ttl = ttl
+
+        ttl_evict
+      end
+
+      def getset(key)
+        ttl_evict
+
+        found = true
+        value = @data_lru.delete(key){ found = false }
+        if found
+          @data_lru[key] = value
+        else
+          result = @data_lru[key] = yield
+          @data_ttl[key] = Time.now.to_f
+
+          if @data_lru.size > @max_size
+            key, _ = @data_lru.first
+
+            @data_ttl.delete(key)
+            @data_lru.delete(key)
+          end
+
+          result
+        end
+      end
+
+      def fetch(key)
+        ttl_evict
+
+        found = true
+        value = @data_lru.delete(key){ found = false }
+        if found
+          @data_lru[key] = value
+        else
+          yield if block_given?
+        end
+      end
+
+      def [](key)
+        ttl_evict
+
+        found = true
+        value = @data_lru.delete(key){ found = false }
+        if found
+          @data_lru[key] = value
+        else
+          nil
+        end
+      end
+
+      def []=(key, val)
+        ttl_evict
+
+        @data_lru.delete(key)
+        @data_ttl.delete(key)
+
+        @data_lru[key] = val
+        @data_ttl[key] = Time.now.to_f
+
+        if @data_lru.size > @max_size
+          key, _ = @data_lru.first
+
+          @data_ttl.delete(key)
+          @data_lru.delete(key)
+        end
+
+        val
+      end
+
+      def each
+        ttl_evict
+
+        array = @data_lru.to_a
+        array.reverse!.each do |pair|
+          yield pair
+        end
+      end
+
+      # used further up the chain, non thread safe each
+      alias_method :each_unsafe, :each
+
+      def to_a
+        ttl_evict
+
+        array = @data_lru.to_a
+        array.reverse!
+      end
+
+      def delete(key)
+        ttl_evict
+
+        @data_lru.delete(key)
+        @data_ttl.delete(key)
+      end
+
+      alias_method :evict, :delete
+
+      def key?(key)
+        ttl_evict
+
+        @data_lru.key?(key)
+      end
+
+      alias_method :has_key?, :key?
+
+      def clear
+        @data_lru.clear
+        @data_ttl.clear
+      end
+
+      def expire
+        ttl_evict
+      end
+
+      def count
+        @data_lru.size
+      end
+
+      # for cache validation only, ensures all is sound
+      def valid?
```

Comment:
I just think "valid?" probably belongs in protected, count and clear belong
in public

do you think valid? belongs in public?

On Mon, Mar 30, 2015 at 11:33 AM, Kaijah Hougham <notifications@github.com>
wrote:

> In lib/lru_redux/ttl/cache.rb
> <https://github.com/SamSaffron/lru_redux/pull/10#discussion_r27361492>:
>
> > +
> > +      def clear
> > +        @data_lru.clear
> > +        @data_ttl.clear
> > +      end
> > +
> > +      def expire
> > +        ttl_evict
> > +      end
> > +
> > +      def count
> > +        @data_lru.size
> > +      end
> > +
> > +      # for cache validation only, ensures all is sound
> > +      def valid?
>
> Which method(s) in particular? #clear, #count and #valid? are part of the
> public methods currently present in the LRU cache and #expire is the
> manual TTL eviction method.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SamSaffron/lru_redux/pull/10/files#r27361492>.
>